sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -1047,6 +391,11 @@ case class ScalaUDF(
     """
   }
 
-  private[this] val converter = CatalystTypeConverters.createToCatalystConverter(dataType)
-  override def eval(input: InternalRow): Any = converter(f(input))
+  lazy val outputEncoder: ExpressionEncoder[Row] =
+      RowEncoder(StructType(StructField("_c0", dataType) :: Nil))
+
+  override def eval(input: InternalRow): Any = {
+    val projected = InternalRow.fromSeq(children.map(_.eval(input)))
+    outputEncoder.toRow(Row(f(projected))).copy().asInstanceOf[InternalRow].get(0, dataType)
```

Comment:
Because `toRow` are to return the same actual `InternalRow` object in `ExpressionEncoder`. We need to copy the result. Otherwise, as I tried, the results will be incorrect.